### DISCUSSION TITLE
How to use variadic functions with `extern(module)`?

### DISCUSSION THREAD

**Discussion by rempas** [MrSmith33/vox#23] [Q&A] [answered]
First of all, I've seen that you implemented it for every platform so great job!!! Now this will make Vox read to be used in production as it can link with any C library! So I'm trying to play with this feature a little bit by trying to use "printf". Now its signature in C is the following: `int printf(const char *format, ...);`. I tried to do it a lot of ways. I even tried to use Vox like templates variadic arguments but nothing works. The latest code I have is:

```
enum u32 stdin  = 0;
enum u32 stdout = 1;
enum u32 stderr = 2;

@extern(syscall, 60)
void exit(i32 error_code);

@extern(syscall, 1)
void sys_write(u32 fd, u8* buf, u64 count);

void write(u32 fd, u8[] data) {
  sys_write(fd, data.ptr, data.length);
}

@extern(module, "libc")
i32 printf(u8*... format);

void main(void* hInstance, void* hPrevInstance, u8* lpCmdLine, i32 nShowCmd) {
  u8[] msg = "hello linux\n";
  write(stdout, msg);
  exit(0);
}
```

Other versions gave me an error for not accepted syntax 
but this version gives me the following error message:
`zsh: illegal hardware instruction (core dumped)  ./vox test.vox`
Any ideas?

**Comment by MrSmith33** [comment, marked as answer]
On linux `@extern(module)` is not useful, as linking shared libraries is still work in progress (I'm close to figuring it out though).

Also, C-style variadic functions are not supported, and I'm not sure that implementing it mostly for `printf` from libc is worth it.

The most complicated and valuable part is float to string conversion though. I haven't yet implemented it using Vox. But here you can see a simple implementation that I use for printing stuff (supports ints, pointers, strings, bools) <https://github.com/MrSmith33/voxelman2/blob/master/plugins/core/utils.vx#L217-L230>

Latest CI builds have debug version, btw

**Reply by rempas** [reply]
I got that! Thanks a lot for your answer!
